### ISSUE TITLE
yolov8 classification model how to get mertics like object detection gives?

### ISSUE THREAD

**Issue by chikyung2** [ultralytics/ultralytics#10117]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have already read this post. https://github.com/ultralytics/yolov5/issues/11509
I recently trained a yolov8n-cls model and want to get the standard metrics for classification tasks, such as accuracy, precision, recall, and F1 score for my "best.pt" model. however, i tried the method in this post but not work.
How can i do it? any code example can explain this. and also anyway to get these metrics for each classes so i can know the accuracy for each classes of the "best.pt"?

**Comment by glenn-jocher** [comment]
Hello! 👋 Great question, and I'm glad to see you've already done some research on this.

For classification tasks with YOLOv8 models like `yolov8n-cls`, you can leverage the `Val` mode to calculate metrics such as accuracy, precision, recall, and the F1 score. Ensure your data configuration (usually `data.yaml`) is correctly set up with validation paths and labels.

Here's a simplified code snippet to guide you through getting metrics for your `best.pt` model:

```python
from ultralytics import YOLO

# Load your trained classification model
model = YOLO('path/to/best.pt')

# Run validation to get metrics - ensure 'data.yaml' is configured correctly
results = model.val(data='data.yaml')
print(results.metrics)
```

For metrics on a per-class basis, the `results` object contains detailed information, including class-specific scores:

```python
# Print class-specific metrics
for class_id, metrics in enumerate(results.metrics.per_class):
    print(f"Class {class_id}:")
    print(f" Accuracy: {metrics['accuracy']}")
    print(f" Precision: {metrics['precision']}")
    print(f" Recall: {metrics['recall']}")
    print(f" F1: {metrics['f1']}")
```

Please ensure that your dataset and configurations are correctly set up to reflect your classes and validation sets. This code should provide a good starting point for evaluating your model's performance on the classification task, including metrics for each class. 

If you have any further questions or run into specific issues, feel free to ask! 🚀

**Comment by chikyung2** [comment]
how should the data.yaml look like? casue i trained a classification model with yolo format, train val test and each classes folder contains images. and i just use model.train(data='./dataset',epochs=10,xx) how should the data.yaml be defined for the classification model? just point to my validation folder?

**Comment by glenn-jocher** [comment]
Hey there! 😊

Looks like there was a little mix-up in the example provided. My apologies for the confusion. For a YOLOv8 classification model, to extract detailed metrics including accuracy for each class directly from the model's output, you may need to dive into the `results_dict` property and parse your metrics from there. Let me quickly guide you to handle this:

Regarding your `data.yaml`, for classification tasks, it should define the paths for your train, val (and optionally, test) datasets along with the list of class names. Something like this:

```yaml
train: ./dataset/train
val: ./dataset/val
# Optional: test: ./dataset/test

nc: 2  # number of classes
names: ['apple', 'orange']
```

For detailed per-class metrics, since the direct method expected does not exist, you would explore the `results_dict` which provides top1 and top5 accuracy globally. Typically, detailed per-class analysis might require custom evaluation scripts where you manually compute these metrics based on predictions and true labels.

However, if you explicitly require per-class metrics like precision, recall, and F1-score, consider validating predictions in a custom script, comparing predictions to your validation dataset ground truth.

For passing validation data using `model.val()`, you can point it to your `data.yaml` that's set up as shown above:

```python
results = model.val(data='data.yaml')
print(results.results_dict)
```

Make sure your `data.yaml` paths correctly point to your data, and your dataset directories are structured accordingly. I hope this clears things up a bit! Let us know if you have any more questions. 🚀

**Comment by chikyung2** [comment]
delete my issue please, your answer is useless

**Comment by glenn-jocher** [comment]
@chikyung2 hi there! I'm sorry to hear that you found the previous response unhelpful. If there's a specific aspect of the issue or additional details you can provide, I'd be glad to assist further. For deleting the issue, GitHub doesn't allow direct deletion of issues, but you can close it if you no longer wish to pursue it. Let us know how we can assist you better! 😊
